crypto/ocsp/ocsp_http.c

```diff
@@ -22,11 +26,335 @@
+        if (i <= 0) {
+            if (BIO_should_retry(rctx->io)) {
+              return -1;
+            }
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+
+        rctx->asn1_len -= i;
+
+        if (rctx->asn1_len > 0) {
+          goto next_io;
+        }
+
+        rctx->state = OHS_ASN1_FLUSH;
+        (void)BIO_reset(rctx->mem);
+
+        OPENSSL_FALLTHROUGH;
+    case OHS_ASN1_FLUSH:
+
+        i = BIO_flush(rctx->io);
+        if (i > 0) {
+            rctx->state = OHS_FIRSTLINE;
+            goto next_io;
+        }
+
+        if (BIO_should_retry(rctx->io)) {
+          return -1;
+        }
+        rctx->state = OHS_ERROR;
+        return 0;
+
+    case OHS_ERROR:
+        return 0;
+
+    case OHS_FIRSTLINE:
+    case OHS_HEADERS:
+
+    // Attempt to read a line in.
+ next_line:
+        // Due to &%^*$" memory BIO behaviour with BIO_gets we have to check
+        // there's a complete line in there before calling BIO_gets or we'll
+        // just get a partial read.
+        n = BIO_get_mem_data(rctx->mem, &p);
+        if ((n <= 0) || !memchr(p, '\n', n)) {
+            if (n >= rctx->iobuflen) {
+                rctx->state = OHS_ERROR;
+                return 0;
+            }
+            goto next_io;
+        }
+        n = BIO_gets(rctx->mem, (char *)rctx->iobuf, rctx->iobuflen);
+
+        if (n <= 0) {
+            if (BIO_should_retry(rctx->mem)) {
+              goto next_io;
+            }
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+
+        // Don't allow excessive lines.
+        if (n == rctx->iobuflen) {
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+
+        // First line.
+        if (rctx->state == OHS_FIRSTLINE) {
+            if (parse_http_line((char *)rctx->iobuf)) {
+                rctx->state = OHS_HEADERS;
+                goto next_line;
+            } else {
+                rctx->state = OHS_ERROR;
+                return 0;
+            }
+        } else {
+            // Look for blank line: end of headers.
+            for (p = rctx->iobuf; *p; p++) {
+                if ((*p != '\r') && (*p != '\n')) {
+                  break;
+                }
+            }
+            if (*p != '\0') {
+              goto next_line;
+            }
+
+            rctx->state = OHS_ASN1_HEADER;
+        }
+
+        OPENSSL_FALLTHROUGH;
+    case OHS_ASN1_HEADER:
+        // Now reading ASN1 header: can read at least 2 bytes which is enough
+        // for ASN1 SEQUENCE header and either length field or at least the
+        // length of the length field.
+        n = BIO_get_mem_data(rctx->mem, &p);
+        if (n < 2) {
+          goto next_io;
+        }
+
+        // Check it is an ASN1 SEQUENCE.
+        if (*p++ != (V_ASN1_SEQUENCE | V_ASN1_CONSTRUCTED)) {
+            rctx->state = OHS_ERROR;
+            return 0;
+        }
+
+        // Check out length field.
+        if ((*p & 0x80) != 0) {
```

Comment:
Worth indicating or using a constant here to make it clear that this expression is checking to see if the length is encoded as long-form (multiple bytes) versus being a length that can be encoded into 7 bits.